### PR TITLE
Bugfix/#1430 catch exception instead of throwable

### PR DESCRIPTION
**Pull request by NovemberChopin**
#### Group 12 Liujinshi M13
**issue:**
[Catch Exception instead of Throwable. #1430](https://github.com/inception-project/inception/issues/1430)

### CODE REVIEW

File: inception-recommendation/src/main/java/de/tudarmstadt/ukp/inception/recommendation/service/RecommendationServiceImpl.java [hunk 1]
```diff
@@ -1009,7 +1010,7 @@ public Predictions computePredictions(User aUser, Project aProject,
 
                         predictions.putPredictions(layer.getId(), suggestions);
                     }
-                    catch (Throwable e) {
+                    catch (RecommendationException | UIMAException | IOException e) {
```
Comment (munterkalmsteiner) [inline inception-recommendation/src/main/java/de/tudarmstadt/ukp/inception/recommendation/service/RecommendationServiceImpl.java]: The reason why a Throwable is used instead of specific exceptions is to prevent that an implementation error in a recommender breaks the service, i.e. prevents other recommenders to be run. I'd say that is intended.

File: inception-recommendation/src/main/java/de/tudarmstadt/ukp/inception/recommendation/tasks/SelectionTask.java [hunk 2]
```diff
@@ -193,7 +195,7 @@ public void run()
                             recommender, user.getUsername(), result,
                             System.currentTimeMillis() - start, activated));
                 }
-                catch (Throwable e) {
+                catch (ConcurrentException | RecommendationException e) {
```
Comment (munterkalmsteiner) [inline inception-recommendation/src/main/java/de/tudarmstadt/ukp/inception/recommendation/tasks/SelectionTask.java]: See this [comment](https://github.com/inception-project/inception/issues/1430#issuecomment-538633971). Catching `Throwable` is intended and by design.

File: inception-recommendation/src/main/java/de/tudarmstadt/ukp/inception/recommendation/tasks/TrainingTask.java [hunk 3]
```diff
@@ -204,7 +202,7 @@ public void run()
                     ctx.close();
                     recommendationService.putContext(user, recommender, ctx);
                 }
-                catch (Throwable e) {
+                catch (RecommendationException | ConcurrentException e) {
```
Comment (munterkalmsteiner) [inline inception-recommendation/src/main/java/de/tudarmstadt/ukp/inception/recommendation/tasks/TrainingTask.java]: Same comment as in `SelectionTask`

File: inception-search-core/src/main/java/de/tudarmstadt/ukp/inception/search/scheduling/TaskConsumer.java [hunk 4]
```diff
@@ -66,7 +67,7 @@ public void run()
                     activeTask.run();
                     log.debug("Indexing task completed: {}", activeTask);
                 }
-                catch (Throwable e) {
+                catch (BeansException e) {
```
Comment (munterkalmsteiner) [inline inception-search-core/src/main/java/de/tudarmstadt/ukp/inception/search/scheduling/TaskConsumer.java]: Same comment as in `Selectiontask`

File: inception-recommendation/src/main/java/de/tudarmstadt/ukp/inception/recommendation/service/RecommendationServiceImpl.java (hunk 1 above)
Comment (reckart) [reply inception-recommendation/src/main/java/de/tudarmstadt/ukp/inception/recommendation/service/RecommendationServiceImpl.java]: It would probably be a good idea to add a comment here (and in similar places) though saying that it is in fact intentional.

File: inception-pdf-editor/src/main/java/de/tudarmstadt/ukp/inception/pdfeditor/pdfextract/PDFExtractor.java [hunk 5]
```diff
@@ -345,7 +345,7 @@ else if (font instanceof PDType0Font) {
         try {
             spaceWidthText = font.getSpaceWidth() * glyphSpaceToTextSpaceFactor;
         }
-        catch (Throwable e) {
+        catch (Exception e) {
```
[the remark this answers is not in the document]
Comment (reckart) [reply inception-pdf-editor/src/main/java/de/tudarmstadt/ukp/inception/pdfeditor/pdfextract/PDFExtractor.java]: It is external code, but we kind of adopted it because the upstream project (https://github.com/yuichiro-s/pdfextract) has been abandoned. So in this case, I'd say the change is in fact an improvement.

File: inception-recommendation/src/main/java/de/tudarmstadt/ukp/inception/recommendation/tasks/SelectionTask.java (hunk 2 above)
Comment (NovemberChopin) [reply inception-recommendation/src/main/java/de/tudarmstadt/ukp/inception/recommendation/tasks/SelectionTask.java]: Dear teacher, what do I need to do next?  Know the reason why a Throwable is used instead of specific exceptions. Close pull request?

Comment (munterkalmsteiner) [reply inception-recommendation/src/main/java/de/tudarmstadt/ukp/inception/recommendation/tasks/SelectionTask.java]: No, you can improve the pull request as follows. Keep the change in `PDFExtractor.java`, this one is good (see [comment](https://github.com/inception-project/inception/pull/1431#discussion_r333212781)). 
For the other files, follow this [comment](https://github.com/inception-project/inception/pull/1431#discussion_r333211582), i.e. keep `Throwable` but add a comment that states something along the lines of: catching `Throwable` is intentional here as we want to continue the execution even if a particular recommender fails.

File: inception-recommendation/src/main/java/de/tudarmstadt/ukp/inception/recommendation/service/RecommendationServiceImpl.java (hunk 1 above)
Comment (NovemberChopin) [reply inception-recommendation/src/main/java/de/tudarmstadt/ukp/inception/recommendation/service/RecommendationServiceImpl.java]: catching Throwable is intentional here as we want to continue the execution even if a particular recommender fails.

File: inception-pdf-editor/src/main/java/de/tudarmstadt/ukp/inception/pdfeditor/pdfextract/PDFExtractor.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (NovemberChopin) [reply inception-pdf-editor/src/main/java/de/tudarmstadt/ukp/inception/pdfeditor/pdfextract/PDFExtractor.java]: Catching Throwable is intentional here as we want to continue the execution even if a particular recommender fails.

File: inception-recommendation/src/main/java/de/tudarmstadt/ukp/inception/recommendation/tasks/SelectionTask.java (hunk 2 above)
Comment (NovemberChopin) [reply inception-recommendation/src/main/java/de/tudarmstadt/ukp/inception/recommendation/tasks/SelectionTask.java]: Catching Throwable is intentional here as we want to continue the execution even if a particular recommender fails.

File: inception-recommendation/src/main/java/de/tudarmstadt/ukp/inception/recommendation/tasks/TrainingTask.java (hunk 3 above)
Comment (NovemberChopin) [reply inception-recommendation/src/main/java/de/tudarmstadt/ukp/inception/recommendation/tasks/TrainingTask.java]: Catching Throwable is intentional here as we want to continue the execution even if a particular recommender fails.

File: inception-search-core/src/main/java/de/tudarmstadt/ukp/inception/search/scheduling/TaskConsumer.java (hunk 4 above)
Comment (NovemberChopin) [reply inception-search-core/src/main/java/de/tudarmstadt/ukp/inception/search/scheduling/TaskConsumer.java]: Catching Throwable is intentional here as we want to continue the execution even if a particular recommender fails.

File: inception-recommendation/src/main/java/de/tudarmstadt/ukp/inception/recommendation/service/RecommendationServiceImpl.java (hunk 1 above)
Comment (munterkalmsteiner) [reply inception-recommendation/src/main/java/de/tudarmstadt/ukp/inception/recommendation/service/RecommendationServiceImpl.java]: :) No, what I meant was adding a comment in the code... The comment must make sense, so adapt as needed.

**Comment by munterkalmsteiner** [comment]
@NovemberChopin commit e17f7fc does not belong into this pull request, right? Please create a new pull request for that feature.
Also, my last [comment](https://github.com/inception-project/inception/pull/1431#discussion_r339208857) in this pull request has not been addressed yet. Do you plan to fix it?

File: inception-recommendation/src/main/java/de/tudarmstadt/ukp/inception/recommendation/service/RecommendationServiceImpl.java (hunk 1 above)
Comment (NovemberChopin) [reply inception-recommendation/src/main/java/de/tudarmstadt/ukp/inception/recommendation/service/RecommendationServiceImpl.java]: Sorry, teacher. I will submit tomorrow

File: inception-recommendation/src/main/java/de/tudarmstadt/ukp/inception/recommendation/service/RecommendationServiceImpl.java [hunk 6]
```diff
@@ -1009,6 +1009,7 @@ public Predictions computePredictions(User aUser, Project aProject,
 
                         predictions.putPredictions(layer.getId(), suggestions);
                     }
+                    // Catching Throwable is intentional here as we want to continue the execution even if a particular recommender fails.
```
Comment (munterkalmsteiner) [inline inception-recommendation/src/main/java/de/tudarmstadt/ukp/inception/recommendation/service/RecommendationServiceImpl.java]: These lines are very long. Please install checkstyle (see [instructions](https://inception-project.github.io//releases/0.12.4/docs/developer-guide.html#_setting_up_checkstyle_and_formatting)) to find style violations in your code.

File: inception-ui-kb/src/main/java/de/tudarmstadt/ukp/inception/ui/kb/ConceptTreePanel.html [hunk 7]
```diff
@@ -37,6 +37,7 @@ <h3 class="panel-title">
     </div>
     <div class="panel-footer text-right">
       <button wicket:id="add" type="submit" class="btn btn-primary"><i class="fa fa-plus" aria-hidden="true"></i> <wicket:container wicket:id="label"></wicket:container></button>
+        <button wicket:id="autoCreateLabel" type="submit" class="btn btn-primary">autoCreateLabel</button>
```
Comment (munterkalmsteiner) [inline inception-ui-kb/src/main/java/de/tudarmstadt/ukp/inception/ui/kb/ConceptTreePanel.html]: This change is not related to this refactoring. Please create a new pull request for feature changes.

File: inception-ui-kb/src/main/java/de/tudarmstadt/ukp/inception/ui/kb/ConceptTreePanel.java [hunk 8]
```diff
@@ -127,13 +127,34 @@ protected boolean isSelected()
         addLink.add(new Label("label", new ResourceModel("concept.list.add")));
         addLink.add(new WriteProtectionBehavior(kbModel));
         add(addLink);
+        /**
```
Comment (munterkalmsteiner) [inline inception-ui-kb/src/main/java/de/tudarmstadt/ukp/inception/ui/kb/ConceptTreePanel.java]: This change is not related to this refactoring. Please create a new pull request for feature changes.

File: inception-ui-kb/src/main/java/de/tudarmstadt/ukp/inception/ui/kb/ConceptTreePanel.java [hunk 9]
```diff
@@ -127,13 +127,34 @@ protected boolean isSelected()
...
     }
-    
+
+    /**
+     * handle click event
+     */
+    protected void autoCreateLabel(AjaxRequestTarget aTarget) throws Exception {
+        List<KBHandle> list = kbService.listAllConcepts(kbModel.getObject(), true);
+        //go through each resource in the KB
+        for(KBHandle kbh : list) {
```
Comment (munterkalmsteiner) [inline inception-ui-kb/src/main/java/de/tudarmstadt/ukp/inception/ui/kb/ConceptTreePanel.java]: This change is not related to this refactoring. Please create a new pull request for feature changes.

File: inception-ui-kb/src/main/java/de/tudarmstadt/ukp/inception/ui/kb/ConceptTreePanel.html (hunk 7 above)
Comment (NovemberChopin) [reply inception-ui-kb/src/main/java/de/tudarmstadt/ukp/inception/ui/kb/ConceptTreePanel.html]: I know, teacher. I have removed the commit e17f7fc from the current pull request

**Comment by reckart** [comment]
Jenkins, can you test this please?
